substrate/tendermint/machine/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+    if let Some(msg) = self.block.message(data) {
+      self.queue.push_back(msg);
+    }
+  }
+
+  // Start a new round. Returns true if we were the proposer
+  fn round(&mut self, round: RoundNumber, time: Option<CanonicalInstant>) -> bool {
+    if let Some(data) =
+      self.block.new_round(round, self.weights.proposer(self.block.number, round), time)
+    {
+      self.broadcast(data);
+      true
+    } else {
+      false
+    }
+  }
+
+  // 53-54
+  async fn reset(&mut self, end_round: RoundNumber, proposal: N::Block) {
+    // Ensure we have the end time data for the last round
+    self.block.populate_end_time(end_round);
+
+    // Sleep until this round ends
+    let round_end = self.block.end_time[&end_round];
+    sleep(round_end.instant().saturating_duration_since(Instant::now())).await;
+
+    // Clear our outbound message queue
+    self.queue = VecDeque::new();
+
+    // Create the new block
+    self.block = BlockData::new(
+      self.weights.clone(),
+      BlockNumber(self.block.number.0 + 1),
+      self.signer.validator_id().await,
+      proposal,
+    );
+
+    // Start the first round
+    self.round(RoundNumber(0), Some(round_end));
+  }
+
+  async fn reset_by_commit(&mut self, commit: Commit<N::SignatureScheme>, proposal: N::Block) {
+    let mut round = self.block.round().number;
+    // If this commit is for a round we don't have, jump up to it
+    while self.block.end_time[&round].canonical() < commit.end_time {
+      round.0 += 1;
+      self.block.populate_end_time(round);
+    }
+    // If this commit is for a prior round, find it
+    while self.block.end_time[&round].canonical() > commit.end_time {
+      if round.0 == 0 {
+        panic!("commit isn't for this machine's next block");
+      }
+      round.0 -= 1;
+    }
+    debug_assert_eq!(self.block.end_time[&round].canonical(), commit.end_time);
+
+    self.reset(round, proposal).await;
+  }
+
+  async fn slash(&mut self, validator: N::ValidatorId) {
+    if !self.block.slashes.contains(&validator) {
+      self.block.slashes.insert(validator);
+      self.network.slash(validator).await;
+    }
+  }
+
+  /// Create a new Tendermint machine, from the specified point, with the specified block as the
+  /// one to propose next. This will return a channel to send messages from the gossip layer and
+  /// the machine itself. The machine should have `run` called from an asynchronous task.
+  #[allow(clippy::new_ret_no_self)]
+  pub async fn new(
+    network: N,
+    last: (BlockNumber, u64),
+    proposal: N::Block,
+  ) -> TendermintHandle<N> {
+    let (msg_send, msg_recv) = mpsc::unbounded();
+    let (step_send, step_recv) = mpsc::unbounded();
+    TendermintHandle {
+      step: step_send,
+      messages: msg_send,
+      machine: {
+        let last_time = sys_time(last.1);
+        // If the last block hasn't ended yet, sleep until it has
+        sleep(last_time.duration_since(SystemTime::now()).unwrap_or(Duration::ZERO)).await;
```

Comment:
This allows a genesis time in the future.

If the current time is 10, and the block is scheduled for 100, this will do 100 - 10 = 90, sleep for 90.

If the current time is 10, and the block occurred at 0, this will do 0 - 10, and hit the exception. ZERO causes a NOP as it shouldn't sleep in that case.